common/metrics/label_maker.go

```diff
@@ -0,0 +1,72 @@
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// labelMaker encapsulates logic for creating labels for metrics.
+type labelMaker struct {
+	keys         []string
+	emptyValues  []string
```

Comment:
If a label is set up with a non-null template, but no labels are provided at runtime, then this `emptyValues` list is passed go prometheus. Prometheus returns an error if you don't pass in the expected number of flags.

```
func (l *labelMaker) extractValues(label any) ([]string, error) {
	// ...

	if label == nil {
		return l.emptyValues, nil
	}
```

We could create a new empty list each time, but I thought it would be more resource efficient to just reuse the same empty list over and over.